crates/sui/src/client_commands.rs

```diff
@@ -458,73 +498,101 @@ impl SuiClientCommands {
         context: &mut WalletContext,
     ) -> Result<SuiClientCommandResult, anyhow::Error> {
         let ret = Ok(match self {
-            SuiClientCommands::Publish {
+            SuiClientCommands::Upgrade {
                 package_path,
-                gas,
+                upgrade_capability,
                 build_config,
+                gas,
                 gas_budget,
                 skip_dependency_verification,
                 with_unpublished_dependencies,
             } => {
                 let sender = context.try_get_object_owner(&gas).await?;
                 let sender = sender.unwrap_or(context.active_address()?);
 
-                let config = resolve_lock_file_path(build_config, Some(package_path.clone()))?;
-                let run_bytecode_verifier = true;
-                let print_diags_to_stderr = true;
-
-                let config = BuildConfig {
-                    config,
-                    run_bytecode_verifier,
-                    print_diags_to_stderr,
-                };
-
-                let resolution_graph = config.resolution_graph(&package_path)?;
-                let dependencies = gather_dependencies(&resolution_graph);
+                let client = context.get_client().await?;
+                let (dependencies, compiled_modules, compiled_package, package_id) =
+                    compile_package(
+                        &client,
+                        build_config,
+                        package_path,
+                        with_unpublished_dependencies,
+                        skip_dependency_verification,
+                    )
+                    .await?;
 
-                check_invalid_dependencies(&dependencies.invalid)?;
+                let SuiObjectResponse::Exists(data) = context
+                    .get_client()
+                    .await?
+                    .read_api()
+                    .get_object_with_options(upgrade_capability, SuiObjectDataOptions::default().with_bcs().with_owner())
```

Comment:
Just noting that it's kind of unfortunate that we have to call RPC twice -- once here to get the digest and policy, and again to get the ownership, but I can't think of a great way to get around that without exposing the notion of `ObjectArg` through the `ProgrammableTransactionBuilder` interface, which we don't do in other cases.